File: leetcode/problems/problem-specific/25.reverse-nodes-in-k-group/25.reverse-nodes-in-k-group.go
package main

// Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

// * Iterative Solution | O(n) Time | O(1) Space

func reverseKGroup(head *ListNode, k int) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	dummyHead := &ListNode{0, head}
	prevGroup := dummyHead
	for {
		kthNode := getKthNode(prevGroup, k)
		if kthNode == nil {
			break
		}

		// * Reverse the nodes in the current group
		nextGroup := kthNode.Next
		prev, cur := kthNode.Next, prevGroup.Next
		for cur != nextGroup {
			next := cur.Next
			cur.Next = prev
			prev = cur
			cur = next
		}

		// * Modify the prevGroup link
		next := prevGroup.Next
		prevGroup.Next = kthNode
		prevGroup = next
	}

	return dummyHead.Next
}

func getKthNode(head *ListNode, k int) *ListNode {
	for head != nil && k > 0 {
		head = head.Next
		k -= 1
	}

	return head
}

File: leetcode/problems/language-specific/python/25.reverse-nodes-in-k-group.py
#
# @lc app=leetcode id=25 lang=python3
#
# [25] Reverse Nodes in k-Group
#

# @lc code=start

# * Iterative Dummy Head Solution | O(n) Time | O(1) Space
# * n -> The number of nodes in the linked list


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not head or not head.next:
            return head

        dummy_head = ListNode(0, head)
        prev_group = dummy_head
        while True:
            kth_node = self.get_kth_node(prev_group, k)
            if not kth_node:
                break

            # * Reverse the linked list in the current group.
            next_group = kth_node.next
            prev, cur = next_group, prev_group.next
            while cur != next_group:
                next = cur.next
                cur.next = prev
                prev = cur
                cur = next

            # * Modify the prev_group link.
            next = prev_group.next
            prev_group.next = kth_node
            prev_group = next

        return dummy_head.next

    def get_kth_node(self, head: Optional[ListNode], k: int):
        while head and k > 0:
            head = head.next
            k -= 1

        return head


# @lc code=end
